lib/minesweeper_2pl/boardcli.rb

```diff
@@ -0,0 +1,150 @@
+    HEADER_CELL_LEFT = "   "
+    HEADER_CELL_RIGHT = "  "
+    INTRO = "     "
+
+    def print_message(string)
+      puts string
+    end
+
+    def board_to_string(board)
+      string = build_header(board)
+      string += header_formatting
+      string += build_rows(board)
+    end
+
+    def header_formatting
+      NEWLINE + ALPHA_LEFT + sprintf("%2s", 0.to_s) + ALPHA_RIGHT
+    end
+
+    def build_header(board)
+      string = NEWLINE + INTRO
+      board.row_size.to_i.times do |count|
+        string += HEADER_CELL_LEFT + sprintf("%2s", (count).to_s) + HEADER_CELL_RIGHT
+      end
+      string += NEWLINE + INTRO
+      board.row_size.to_i.times do
+        string += CELL_DIVIDER
+      end
+      string += "+"
+    end
+
+    def build_rows(board)
+      string = ""
+      (0..board.row_size.to_i-1).to_a.each do |row|
+        string += build_cell(board, row)
+        string += build_row_divider(board, row)
+      end
+      string
+    end
+
+    def build_cell(board, row)
+      string = ""
+      (0..board.row_size.to_i-1).to_a.each do |col|
+        if col == board.row_size - 1
+          string += CELL_LEFT + get_cell_content(board.positions, (row*board.row_size.to_i)+(col)) + CELL_END
+        else
+          string += CELL_LEFT + get_cell_content(board.positions, (row*board.row_size.to_i)+(col)) + CELL_RIGHT
+        end
+      end
+      string
+    end
+
+    def build_row_divider(board, row)
+      string = NEWLINE + INTRO
+      board.row_size.to_i.times do
+        string += CELL_DIVIDER
+      end
+      string += "+"
+      unless row == board.row_size - 1
+        string += NEWLINE + ALPHA_LEFT + sprintf("%2s", (row+1).to_s) + ALPHA_RIGHT
+      end
+      string
+    end
+
+    def get_cell_content(positions, cell)
+      cell = positions[cell]
+      if show_bombs == "show"
+        render_lost_view(cell)
+      elsif show_bombs == "won"
+        render_won_view(cell)
+      else
+        render_normal_view(cell)
+      end
+    end
+
+    private
+
+    def render_lost_view(cell)
+      if cell_is_a_bomb_or_flag?(cell)
+        show_bomb_emoji
+      elsif cell_is_a_bomb?(cell)
+        show_bomb_emoji
+      elsif cell_is_a_flag?(cell)
+        show_flag_emoji
+      elsif cell.length == 1
+        cell + " "
+      end
+    end
+
+    def render_won_view(cell)
+      if cell_is_a_bomb_or_flag?(cell)
+        show_trophy_emoji
+      elsif cell_is_a_flag?(cell)
+        show_flag_emoji
+      elsif cell.length == 1
+        cell + " "
+      end
+    end
+
+    def render_normal_view(cell)
+      if cell_is_a_flag?(cell)
+        show_flag_emoji
+      elsif cell_is_a_bomb?(cell)
+        show_empty
+      elsif cell.length == 1
+        cell + " "
+      end
+    end
+
+    def cell_is_a_bomb?(cell)
+      cell.include? "B"
+    end
+
+    def cell_is_a_flag?(cell)
+      cell.include? "F"
+    end
+
+    def cell_is_a_bomb_or_flag?(cell)
+      cell.include? "BF"
+    end
+
+    def show_bomb_emoji
```

Comment:
Here's another place where thinking carefully about return types can inform good naming decisions. This method just returns a string—it's really just a value. But it has a verb ("show") in its name. This method doesn't actually "show" anything—it just returns a string as a value, and it's up to the caller to print it out, hand it off to some other object, or do whatever. Since this is Just A Value, it might be a good idea to remove the verb from the name.

Reply:
...and once it's Just A Value, it might be better represented by a constant, or e.g. an `attr_getter` that wraps a constant.